src/splay_tree.jl

```diff
@@ -126,7 +126,7 @@ function _join!(tree::SplayTree, s::Union{SplayTreeNode, Nothing}, t::Union{Spla
     end
 end
 
-function search_node(tree::SplayTree{K}, d::K) where K
+function search_node!(tree::SplayTree{K}, d::K) where K
```

Comment:
I gave this a thought. And perhaps now I recall the reason why I didn't splay while searching. 

Making `search_node` as a modifying function essentially means we are modifying the tree on every access. This means simple lookup using `in` or `haskey` - all such functions need to be appended with `!` indicating the change in stucture of the underlying tree. This will be a breaking change (and quite rightly so).

The context for use of splay trees implies you want recently accessed elements near to the root, so that the subsequent access takes less time. 

@oxinabox are we okay for introducing a breaking change in the package?

Reply:
Yes into the master branch that is fine. We are still prepping 0.19

Reply:
I tend to agree with @fingolfin here: none of these operations actually modify the tree in any way that's visible with user-facing set operations. Even an in-order traversal would not change after a splay (though pre/post-orders would, but I don't think those are exposed?).

I've pushed up a new version that reverts this change and leaves `search_node` un-`!`d. Thoughts on this direction @eulerkochy ? Or do you strongly prefer changing all of the accessors to `!`?

Reply:
I am happy with the un-`!` with being defined as something not changing the "user visible state". Let's  add a comment to that effect, so that it's remember why this decision was made, even though the "internal state" of the tree is changing.